### PR TITLE
Flag persistent WalSndError on repslot invalidation

### PR DESCRIPTION
**Pull request by soumyadeep2007**
67d4843a149 introduced WalSndError as a means for having a persistent
error state, exposed through the sync_error field in
gp_stat_replication. This state was meant to signal
walsender<->walreceiver connection failures, even when the walsender
died (the traditional upstream mechanism depends on the walsender being
alive to report issues).

We should set this state when the replication slot is invalidated as
well. This was done correctly in 6X_STABLE commit: ea69506b407.

This would prevent flakes in the associated max_slot_wal_keep_size test
such as:

```diff
--- expected/segwalrep/max_slot_wal_keep_size.out
+++ results/segwalrep/max_slot_wal_keep_size.out
@@ -289,7 +450,7 @@
 1: SELECT sync_error FROM gp_stat_replication WHERE gp_segment_id = 0;
  sync_error
 ------------
- walread
+ none
 (1 row)
```

PS: The max_slot_wal_keep_size GUC was first backported to 6X and then
later backported to 7X from PG13.

Co-authored-by: Alexandra Wang <walexandra@vmware.com>

Dev-pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/fix_max_slot_flake

### CODE REVIEW

File: src/backend/replication/slot.c [hunk 1]
```diff
@@ -1337,6 +1338,8 @@ InvalidateObsoleteReplicationSlots(XLogSegNo oldestSegno)
 			goto restart;
 		}
 	}
+	/* GPDB: Set WalSndCtl state to indicate persistent sync error state */
+	WalSndCtl->error = WALSNDERROR_WALREAD;
```
Comment (huansong) [inline src/backend/replication/slot.c:1342]: In case no replication slot is invalidated, should we still change this value to `WALREAD`? Should we instead place this line inside the `if (invalidated)` block later so that only when something's wrong we would set the state?

**Review by huansong — APPROVED** [review]
Fix LGTM. I think there's an issue with the place where we set `WALSNDERROR_WALREAD`. 

Also, curious about about why before this PR the test `max_slot_wal_keep_size` was just flaky instead of always printing `none`.  Do you know where did we set `WALREAD` when the test passed?

**Comment by soumyadeep2007** [comment]
> Also, curious about about why before this PR the test max_slot_wal_keep_size was just flaky instead of always printing none. Do you know where did we set WALREAD when the test passed?

The only places we set the "walread" error is in `XLogRead`, when the walsender tries to read a wal segment and it can't (open/lseek/read).

We also clear the error on a successful wal read at the end of `XLogRead`.

My guess is that when we do the checkpoint and the replication slot is invalidated, it could be one of:
(i)   the walsender has attempted to read a wal that doesn't exist as it was removed by the CHECKPOINT
(ii)  the walsender has not attempted to read any wal yet after its last successful attempt (none would have been set and that remains)
Its a timing issue I think. Now by setting this explicitly, I "think" this flake will go away. or at least manifest less frequently.

**Comment by huansong** [comment]
> (i) the walsender has attempted to read a wal that doesn't exist as it was removed by the CHECKPOINT
(ii) the walsender has not attempted to read any wal yet after its last successful attempt (none would have been set and that remains)
Its a timing issue I think. Now by setting this explicitly, I "think" this flake will go away

Thanks for the explanation. Makes sense. Yeah in any case the CHECKPOINT should always set it now.
